### PR TITLE
[SYCL-MLIR] Improve code generation for the rest of binary operators

### PR DESCRIPTION
**Pull request by victor-eds**
9ac759f12e55 improved code generation for addition and subtraction. This commit addresses the rest of binary arithmetic operators (mul, div, rem, shl, shr, and, or, xor):

- Enable code generation for mul and div operators and vector operands;
- Mask shl and shr operands in OpenCL
- Always generate signed/unsigned operations for vector operands.

Signed-off-by: Victor Perez <victor.perez@codeplay.com>

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 1]
```diff
@@ -2114,18 +2115,6 @@ ValueCategory MLIRScanner::VisitBinAssign(BinaryOperator *e) {
...
-  // TODO note assumptions made here about unsigned / unordered
-  bool signedType = true;
-  if (auto bit = dyn_cast<clang::BuiltinType>(Ty)) {
-    if (bit->isUnsignedInteger())
-      signedType = false;
-    if (bit->isSignedInteger())
-      signedType = true;
-  }
-  return signedType;
-}
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2127]: Not needed anymore: used by former code

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 2]
```diff
@@ -2848,56 +2853,99 @@ ValueCategory MLIRScanner::EmitBinSub(const BinOpInfo &Info) {
...
 ValueCategory MLIRScanner::EmitBinOr(const BinOpInfo &Info) {
-  // TODO short circuit
-  return ValueCategory(builder.create<OrIOp>(loc,
-                                             Info.getLHS().getValue(builder),
-                                             Info.getRHS().getValue(builder)),
-                       /*isReference*/ false);
+  const auto Loc = getMLIRLocation(Info.getExpr()->getExprLoc());
+  auto LHS = Info.getLHS();
+  auto RHS = Info.getRHS();
+  return LHS.Or(builder, Loc, RHS.val);
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2950]: No change in these operators. Just using new code with asserts.

File: polygeist/tools/cgeist/Lib/clang-mlir.h [hunk 3]
```diff
@@ -523,6 +523,8 @@ class MLIRScanner : public clang::StmtVisitor<MLIRScanner, ValueCategory> {
 #include "Expressions.def"
 #undef HANDLEBINOP
 
+  ValueCategory ConstrainShiftValue(ValueCategory LHS, ValueCategory RHS);
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/clang-mlir.h:526]: Needed in OpenCL.

File: polygeist/tools/cgeist/Test/Verification/and.c [hunk 4]
```diff
@@ -0,0 +1,190 @@
+// RUN: cgeist -O0 %s --function=* -S | FileCheck %s
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/and.c:1]: No changes here.

File: polygeist/tools/cgeist/Test/Verification/div.c [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+// CHECK-SAME:                        %[[VAL_0:.*]]: memref<?xvector<3xf64>>,
+// CHECK-SAME:                        %[[VAL_1:.*]]: memref<?xvector<3xf64>>) -> vector<3xf64>
+// CHECK-NEXT:      %[[VAL_2:.*]] = affine.load %[[VAL_0]][0] : memref<?xvector<3xf64>>
+// CHECK-NEXT:      %[[VAL_3:.*]] = affine.load %[[VAL_1]][0] : memref<?xvector<3xf64>>
+// CHECK-NEXT:      %[[VAL_4:.*]] = arith.divf %[[VAL_2]], %[[VAL_3]] : vector<3xf64>
+// CHECK-NEXT:      return %[[VAL_4]] : vector<3xf64>
+// CHECK-NEXT:    }
+double_vec div_vf64(double_vec a, double_vec b) {
+  return a / b;
+}
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/div.c:250]: Was not working

File: polygeist/tools/cgeist/Test/Verification/mul.c [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+// CHECK-SAME:                        %[[VAL_0:.*]]: memref<?xvector<3xf64>>,
+// CHECK-SAME:                        %[[VAL_1:.*]]: memref<?xvector<3xf64>>) -> vector<3xf64>
+// CHECK-NEXT:      %[[VAL_2:.*]] = affine.load %[[VAL_0]][0] : memref<?xvector<3xf64>>
+// CHECK-NEXT:      %[[VAL_3:.*]] = affine.load %[[VAL_1]][0] : memref<?xvector<3xf64>>
+// CHECK-NEXT:      %[[VAL_4:.*]] = arith.mulf %[[VAL_2]], %[[VAL_3]] : vector<3xf64>
+// CHECK-NEXT:      return %[[VAL_4]] : vector<3xf64>
+// CHECK-NEXT:    }
+double_vec mul_vf64(double_vec a, double_vec b) {
+  return a * b;
+}
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/mul.c:150]: Working now

File: polygeist/tools/cgeist/Test/Verification/or.c [hunk 7]
```diff
@@ -0,0 +1,189 @@
+// RUN: cgeist -O0 %s --function=* -S | FileCheck %s
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/or.c:1]: No changes for this operand.

File: polygeist/tools/cgeist/Test/Verification/shl.cl [hunk 8]
```diff
@@ -0,0 +1,234 @@
+// RUN: cgeist -O0 %s --function=* -S | FileCheck %s
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/shl.cl:1]: Applies OpenCL masking now.

File: polygeist/tools/cgeist/Test/Verification/shr.cl [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+// CHECK-NEXT:      %[[VAL_3:.*]] = arith.extui %[[VAL_0]] : i8 to i32
+// CHECK-NEXT:      %[[VAL_4:.*]] = arith.extui %[[VAL_1]] : i8 to i32
+// CHECK-NEXT:      %[[VAL_5:.*]] = arith.andi %[[VAL_4]], %[[VAL_2]] : i32
+// CHECK-NEXT:      %[[VAL_6:.*]] = arith.shrsi %[[VAL_3]], %[[VAL_5]] : i32
+// CHECK-NEXT:      %[[VAL_7:.*]] = arith.trunci %[[VAL_6]] : i32 to i8
+// CHECK-NEXT:      return %[[VAL_7]] : i8
+// CHECK-NEXT:    }
+unsigned char shr_ui8(unsigned char a, unsigned char b) {
+  return a >> b;
+}
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Test/Verification/shr.cl:84]: Operations with unsigned operands needing an extension are signed, as the extension is to signed `int`.

**Comment by victor-eds** [comment]
Almost all of the new lines are tests.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 10]
```diff
@@ -2848,56 +2853,99 @@ ValueCategory MLIRScanner::EmitBinSub(const BinOpInfo &Info) {
   return DiffInChars.ExactSDiv(builder, Loc, Divisor);
 }
 
+static ValueCategory GetWidthMinusOneValue(mlir::OpBuilder &Builder,
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2856]: Why not return a Value?

Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc]: True. Done.

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/tools/cgeist/Lib/ValueCategory.cc [hunk 11]
```diff
@@ -293,6 +293,13 @@ template <typename OpTy> inline void warnUnconstrainedOp() {
       << "Creating unconstrained " << OpTy::getOperationName() << "\n";
 }
 
+template <typename OpTy> inline void warnNonExactOp(bool IsExact) {
+  if (!IsExact)
+    return;
+  llvm::WithColor::warning()
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/ValueCategory.cc:299]: mlirclang::warning()

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 12]
```diff
@@ -2594,47 +2583,63 @@ static void informNoOverflowCheck(LangOptions::SignedOverflowBehaviorTy SOB,
...
 ValueCategory MLIRScanner::EmitBinMul(const BinOpInfo &Info) {
-  auto lhs_v = Info.getLHS().getValue(builder);
-  auto rhs_v = Info.getRHS().getValue(builder);
-  if (lhs_v.getType().isa<mlir::FloatType>()) {
-    return ValueCategory(builder.create<arith::MulFOp>(loc, lhs_v, rhs_v),
-                         /*isReference*/ false);
-  } else {
-    return ValueCategory(builder.create<arith::MulIOp>(loc, lhs_v, rhs_v),
-                         /*isReference*/ false);
+  const auto Loc = getMLIRLocation(Info.getExpr()->getExprLoc());
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:2563]: I'd say these three values are pretty transparent; `auto` works here (as in the rest of functions).

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 13]
```diff
@@ -2848,56 +2853,99 @@ ValueCategory MLIRScanner::EmitBinSub(const BinOpInfo &Info) {
...
+  auto LHS = Info.getLHS();
+  auto RHS = Info.getRHS();
+
+  // LLVM requires the LHS and RHS to be the same type: promote or truncate the
+  // RHS to the same size as the LHS.
+  if (LHS.val.getType() != RHS.val.getType())
+    RHS = RHS.IntCast(builder, Loc, LHS.val.getType(), /*IsSigned*/ false);
+
+  if (Glob.getCGM().getLangOpts().OpenCL) {
+    loc = Loc;
```
[the remark this answers is not in the document]
Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc]: Yep, so that it can be used when in `ConstrainShiftValue`

**Review by etiotto — APPROVED** [review]

**Pull request merged**